### PR TITLE
[FIXED JENKINS-27289] Make MaskingClassLoader work on slaves

### PR DESCRIPTION
**Pull request by liorhson**
Masking classLoader doesn`t work on remote Slave

After upgrading the Apache **httpClient** to 4.4, I found that other plugins such as Git-Client/Maven-plugin that also uses httpClient with different versions collapses with ours. So After I added the Mask-Classes attribute the problem was gone, but it looks like this isolated environment dose not apply for Slaves.
I found out that in the `RemoteClassLoader` while it tries to find a specific class URL, `Which.classFileUrl(class)` uses the `getResource` method. This method is not override by the `hudson.util.MaskingClassLoader`. So the class comes from the right source under the plugin itself, but the resource URL of the same class can came from other source such as other plugins that contain the same classes.

Link to Jira: https://issues.jenkins-ci.org/browse/JENKINS-27289

### CODE REVIEW

**Comment by daniel-beck** [comment]
Why are you adding a second collection with essentially the same contents rather than reuse the `masks` list for resources as well?

**Comment by liorhson** [comment]
Well, the path of a resource is with `'/'`, and class is with `'.'`, so I wanted to store it in a different list to save the String `replace` operation for every resource that the classLoader will look for

**Comment by liorhson** [comment]
Hey, 
Any ETA about this pull request?

File: core/src/main/java/hudson/util/MaskingClassLoader.java [hunk 1]
```diff
@@ -41,7 +42,9 @@
     /**
      * Prefix of the packages that should be hidden.
      */
-    private List<String> masks;
+    private List<String> masksClasses;
+
+    private List<String> masksResources;
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/MaskingClassLoader.java]: minor: items could be made final

File: test/src/test/java/hudson/ClassicPluginStrategyTest.java [hunk 2]
```diff
@@ -87,4 +88,19 @@ public void testDisabledDependencyClassLoader() throws Exception {
...
+     * "foo1" plugin contains attribute of Mask-Classes: org.apache.http.
+     */
+    @LocalData
+    @Issue("JENKINS-27289")
+    public void testMaskResourceClassLoader() throws Exception {
+        PluginWrapper pw = jenkins.getPluginManager().getPlugin("foo1");
+        Class<?> clazz = pw.classLoader.loadClass("org.apache.http.impl.io.SocketInputBuffer");
+        ClassLoader cl = clazz.getClassLoader();
+        URL url = cl.getResource("org/apache/http/impl/io/SocketInputBuffer.class");
+        assert url != null;
```
Comment (oleg-nenashev) [inline test/src/test/java/hudson/ClassicPluginStrategyTest.java]: Minor: Its better to use JUnit assertions

**Comment by oleg-nenashev** [comment]
LGTM in general

**Comment by eyalbe4** [comment]
Hi @oleg-nenashev ,
We are waiting for this pull request to be accepted, so that we can add SNI support to the Artifactory Plugin. If you don't have more comment or questions, can you please merge it?
Thanks for your help!

File: core/src/main/java/hudson/util/MaskingClassLoader.java [hunk 3]
```diff
@@ -62,7 +72,20 @@ public MaskingClassLoader(ClassLoader parent, Collection<String> masks) {
         return super.loadClass(name, resolve);
     }
 
+    @Override
+    public synchronized URL getResource(String name) {
```
Comment (jglick) [inline core/src/main/java/hudson/util/MaskingClassLoader.java]: You need to override `getResources` too. Maybe not for this particular purpose, but for other callers.

**Comment by jglick** [comment]
In the interests of expediency I will merge this as is, but leave the issue open until `getResources` is fixed as well.

**Comment by oleg-nenashev** [comment]
"Merge branch 'master' into JENKINS-27289" flooded the commits history BTW

File: core/src/main/java/hudson/util/MaskingClassLoader.java (hunk 3 above)
Comment (liorhson) [inline core/src/main/java/hudson/util/MaskingClassLoader.java]: Thanks for the merge, where should I add the fix for the `getResources` ? Is it in a new pull request?

**Comment by jglick** [comment]
@liorhson

> where should I add the fix for the `getResources`? Is it in a new pull request?

Yes, if you want to work on that just file a new PR. Thanks!

**Comment by liorhson** [comment]
BTW, in the last two LTS releases, #1598 was not added, although it is already merged, and his reference Jira ticket (JENKINS-27289) is labeled as **lts-candidate**.
Any news about that?

**Comment by daniel-beck** [comment]
@liorhson Please see the [wiki](https://wiki.jenkins-ci.org/display/JENKINS/LTS+Release+Line) for the LTS process. This change did not yet qualify for 1.609.1.

**Comment by liorhson** [comment]
Hey @jglick 
Any particular reason not to **Resolve** Jira ticket JENKINS-27289?

This issue is a real blocker for Jenkins Artifactory plugin (Jfrog)

**Comment by daniel-beck** [comment]
@liorhson Jesse wrote [in a comment](https://issues.jenkins-ci.org/browse/JENKINS-27289?focusedCommentId=228160&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-228160) that the fix is incomplete.

**Comment by daniel-beck** [comment]
@liorhson I suggested [on the dev list](https://groups.google.com/d/msg/jenkinsci-dev/LZ3qWDuJ2IU/sJcn9VXfEnUJ) that this could be included in 1.609.2 despite the associated issue being unresolved.

**Pull request merged**